src/feature/featuremodel.cpp

```diff
@@ -157,144 +169,152 @@ QObject* FeatureModel::get(const QString& feature) {
+      logger.error() << "Error in the json format: experimentalFeature" << key
+                     << "is not an object.";
+      experimentalFeaturesToToggleOn.removeAll(key);
       continue;
     }
 
-    AdjustFiltering::instance()->denyField(key, value.toString());
-  }
+    QJsonObject experimentalFeatureSettingsObj =
+        experimentalFeatureSettings.toObject();
+    for (const QString& settingKey : experimentalFeatureSettingsObj.keys()) {
+      auto value = experimentalFeatureSettingsObj[settingKey].toVariant();
+      if (!value.isValid() || value.isNull()) {
+        logger.warning()
+            << "Received null value for experimental feature setting"
+            << settingKey;
+        continue;
+      }
 
-  QJsonValue mirrorParameterValue = adjustFieldsValue["mirrorParameters"];
-  if (!mirrorParameterValue.isObject()) {
-    logger.error()
-        << "Error in the json format; mirror parameters are not an object";
-    return;
-  }
+      logger.debug() << "Setting experimental feature setting" << settingKey;
 
-  QJsonObject mirrorParameterObject = mirrorParameterValue.toObject();
-  for (const QString& key : mirrorParameterObject.keys()) {
-    QJsonValue values = mirrorParameterObject.value(key);
-    if (!values.isArray()) {
-      logger.error() << "Error in the json format; mirror parameters value is "
-                        "not an array";
-      continue;
+      experimentalFeature->settingGroup()->set(settingKey, value);
     }
 
-    QJsonArray valuesArray = values.toArray();
-    if (valuesArray.size() != 2) {
-      logger.error()
-          << "Error in the json format; mirror value is not an array";
-      continue;
-    }
+    experimentalFeaturesToToggleOff.removeAll(key);
+  }
 
-    QJsonValue mirrorParamValue = valuesArray.first();
-    if (!mirrorParamValue.isString()) {
-      logger.error()
-          << "Error in the json format; mirroring field is not a string";
-      continue;
-    }
+  result.first = experimentalFeaturesToToggleOn;
+  result.second = experimentalFeaturesToToggleOff;
+  return result;
+}
 
-    QJsonValue defaultValue = valuesArray.last();
-    if (!defaultValue.isString()) {
-      logger.error()
-          << "Error in the json format; mirror default value is not a string";
-      continue;
+void FeatureModel::updateFeatureList(const QByteArray& data) {
+  QJsonObject json = QJsonDocument::fromJson(data).object();
+
+  SettingsHolder* settingsHolder = SettingsHolder::instance();
+
+  QStringList featuresToToggleOn = settingsHolder->featuresFlippedOn();
+  QStringList featuresToToggleOff = settingsHolder->featuresFlippedOff();
+
+  if (json.contains("featuresOverwrite")) {
+    QJsonValue features = json["featuresOverwrite"];
+    auto parsedFeatures = parseFeatures(features);
+
+    featuresToToggleOn = parsedFeatures.first;
+    featuresToToggleOff = parsedFeatures.second;
+  }
+
+  if (json.contains("experimentalFeatures")) {
```

Comment:
If the payload contains "featuresOverwrite" and "experimentalFeatures", this approach could break because featuresToToggleOn & featuresToToggleOff are reinitialized on lines 292 & 293 to only the features in "featuresOverwrite"

Reply:
Yeah I know. I have left that to be addressed in a follow up https://mozilla-hub.atlassian.net/browse/VPN-6078.